#VRML_SIM R2023a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/road/protos/Road.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/road/protos/RoadIntersection.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/road/protos/Roundabout.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/road/protos/StraightRoadSegment.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/buildings/protos/SimpleBuilding.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/traffic/protos/TrafficLight.proto"

WorldInfo {
  title "Autonomous Driving Simulation"
  basicTimeStep 32
  contactProperties [
    ContactProperties {
      material1 "WheelMaterial"
      material2 "RoadMaterial"
      coulombFriction [
        0.9
      ]
    }
  ]
}

Viewpoint {
  orientation -0.24 0.91 0.34 1.8
  position 5.2 12.8 8.9
  follow "AGENT_CAR"
}

TexturedBackground {
}

TexturedBackgroundLight {
}

# Main road network
Road {
  translation 0 0 0
  rotation 0 1 0 0
  width 8
  numberOfLanes 2
  rightBorder FALSE
  leftBorder FALSE
  wayPoints [
    0 0 0
    20 0 0
    40 0 0
  ]
}

Road {
  translation 0 0 0
  rotation 0 1 0 1.5708
  width 8
  numberOfLanes 2
  rightBorder FALSE
  leftBorder FALSE
  wayPoints [
    0 0 0
    0 0 20
    0 0 40
  ]
}

# Intersection
RoadIntersection {
  translation 20 0 20
  rotation 0 1 0 0
  id "intersection_1"
  shape [
    0 0 0
    20 0 0
    20 0 20
    0 0 20
  ]
}

# Roundabout
Roundabout {
  translation -10 0 15
  rotation 0 1 0 0
  id "roundabout_1"
  radius 8
  subdivision 16
}

# Parking area
StraightRoadSegment {
  translation 8 0 5
  rotation 0 1 0 1.5708
  width 6
  length 10
  rightBorder FALSE
  leftBorder FALSE
}

# Buildings for visual reference
SimpleBuilding {
  translation 30 0 30
  floorNumber 3
  width 8
  depth 8
  height 12
}

SimpleBuilding {
  translation -20 0 30
  floorNumber 2
  width 6
  depth 10
  height 8
}

SimpleBuilding {
  translation 30 0 -10
  floorNumber 4
  width 10
  depth 6
  height 16
}

# Traffic lights
DEF TRAFFIC_LIGHT_1 TrafficLight {
  translation 18 0 22
  rotation 0 1 0 -1.5708
  startGreen FALSE
}

DEF TRAFFIC_LIGHT_2 TrafficLight {
  translation 22 0 18
  rotation 0 1 0 0
  startGreen FALSE
}

DEF TRAFFIC_LIGHT_3 TrafficLight {
  translation 22 0 22
  rotation 0 1 0 1.5708
  startGreen FALSE
}

# Supervisor robot for environment control
DEF SUPERVISOR Robot {
  controller "supervisor_controller"
  supervisor TRUE
  children [
    Display {
      width 200
      height 100
    }
  ]
}

# Agent car with sensors
DEF AGENT_CAR Robot {
  translation 2 0.1 2
  rotation 0 1 0 0
  children [
    # Car body
    DEF CAR_BODY Shape {
      appearance Appearance {
        material Material {
          diffuseColor 0.1 0.4 0.8
          emissiveColor 0.05 0.2 0.4
        }
      }
      geometry Box {
        size 2 1 4
      }
    }
    
    # Front camera
    DEF CAMERA_MOUNT Transform {
      translation 0 0.3 1.8
      children [
        Camera {
          name "camera"
          fieldOfView 1.2
          width 128
          height 128
          near 0.1
          far 50
        }
      ]
    }
    
    # Distance sensors
    DEF FRONT_SENSOR_MOUNT Transform {
      translation 0 0 2
      children [
        DistanceSensor {
          name "front_sensor"
          lookupTable [
            0 0 0
            1 1000 0
            2 2000 0
          ]
          type "laser"
          numberOfRays 1
          aperture 0.1
        }
      ]
    }
    
    DEF REAR_SENSOR_MOUNT Transform {
      translation 0 0 -2
      rotation 0 1 0 3.14159
      children [
        DistanceSensor {
          name "rear_sensor"
          lookupTable [
            0 0 0
            1 1000 0
            2 2000 0
          ]
          type "laser"
          numberOfRays 1
          aperture 0.1
        }
      ]
    }
    
    DEF LEFT_SENSOR_MOUNT Transform {
      translation -1 0 0
      rotation 0 1 0 -1.5708
      children [
        DistanceSensor {
          name "left_sensor"
          lookupTable [
            0 0 0
            1 1000 0
            2 2000 0
          ]
          type "laser"
          numberOfRays 1
          aperture 0.1
        }
      ]
    }
    
    DEF RIGHT_SENSOR_MOUNT Transform {
      translation 1 0 0
      rotation 0 1 0 1.5708
      children [
        DistanceSensor {
          name "right_sensor"
          lookupTable [
            0 0 0
            1 1000 0
            2 2000 0
          ]
          type "laser"
          numberOfRays 1
          aperture 0.1
        }
      ]
    }
    
    # Inertial measurement unit
    InertialUnit {
      name "inertial_unit"
    }
    
    # Wheels and motors
    DEF FRONT_LEFT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        position 0
        axis 0 1 0
        anchor -0.8 0 1.2
      }
      device [
        RotationalMotor {
          name "front_left_motor"
          maxVelocity 10
        }
      ]
      endPoint Solid {
        translation -0.8 0 1.2
        rotation 1 0 0 1.5708
        children [
          DEF WHEEL Shape {
            appearance Appearance {
              material Material {
                diffuseColor 0.2 0.2 0.2
              }
            }
            geometry Cylinder {
              height 0.3
              radius 0.4
            }
          }
        ]
        contactMaterial "WheelMaterial"
        boundingObject USE WHEEL
        physics Physics {
          density -1
          mass 10
        }
      }
    }
    
    DEF FRONT_RIGHT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        position 0
        axis 0 1 0
        anchor 0.8 0 1.2
      }
      device [
        RotationalMotor {
          name "front_right_motor"
          maxVelocity 10
        }
      ]
      endPoint Solid {
        translation 0.8 0 1.2
        rotation 1 0 0 1.5708
        children [
          USE WHEEL
        ]
        contactMaterial "WheelMaterial"
        boundingObject USE WHEEL
        physics Physics {
          density -1
          mass 10
        }
      }
    }
    
    DEF REAR_LEFT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        position 0
        axis 0 1 0
        anchor -0.8 0 -1.2
      }
      device [
        RotationalMotor {
          name "rear_left_motor"
          maxVelocity 10
        }
      ]
      endPoint Solid {
        translation -0.8 0 -1.2
        rotation 1 0 0 1.5708
        children [
          USE WHEEL
        ]
        contactMaterial "WheelMaterial"
        boundingObject USE WHEEL
        physics Physics {
          density -1
          mass 10
        }
      }
    }
    
    DEF REAR_RIGHT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        position 0
        axis 0 1 0
        anchor 0.8 0 -1.2
      }
      device [
        RotationalMotor {
          name "rear_right_motor"
          maxVelocity 10
        }
      ]
      endPoint Solid {
        translation 0.8 0 -1.2
        rotation 1 0 0 1.5708
        children [
          USE WHEEL
        ]
        contactMaterial "WheelMaterial"
        boundingObject USE WHEEL
        physics Physics {
          density -1
          mass 10
        }
      }
    }
    
    # Steering mechanism (simplified)
    DEF STEERING_JOINT HingeJoint {
      jointParameters HingeJointParameters {
        position 0
        axis 0 1 0
        anchor 0 0 1.2
      }
      device [
        RotationalMotor {
          name "steering_motor"
          maxVelocity 2
          maxTorque 50
        }
      ]
      endPoint Solid {
        translation 0 0 1.2
        children [
          Shape {
            appearance Appearance {
              material Material {
                diffuseColor 0.5 0.5 0.5
              }
            }
            geometry Box {
              size 0.1 0.1 0.1
            }
          }
        ]
        physics Physics {
          density -1
          mass 1
        }
      }
    }
    
    # Keyboard for manual control
    Keyboard {
    }
  ]
  name "AGENT_CAR"
  controller "agent_driver"
  contactMaterial "WheelMaterial"
  boundingObject USE CAR_BODY
  physics Physics {
    density -1
    mass 1000
    centerOfMass [
      0 0 0
    ]
  }
}

# Additional static obstacles for more complex scenarios
Solid {
  translation 15 0.5 8
  children [
    Shape {
      appearance Appearance {
        material Material {
          diffuseColor 0.8 0.2 0.2
        }
      }
      geometry Box {
        size 2 1 1
      }
    }
  ]
  name "obstacle_1"
  boundingObject Box {
    size 2 1 1
  }
}

Solid {
  translation 25 0.5 25
  children [
    Shape {
      appearance Appearance {
        material Material {
          diffuseColor 0.2 0.8 0.2
        }
      }
      geometry Box {
        size 1.5 1 3
      }
    }
  ]
  name "obstacle_2"
  boundingObject Box {
    size 1.5 1 3
  }
}

Solid {
  translation -5 0.5 10
  children [
    Shape {
      appearance Appearance {
        material Material {
          diffuseColor 0.8 0.8 0.2
        }
      }
      geometry Cylinder {
        height 1
        radius 0.5
      }
    }
  ]
  name "obstacle_3"
  boundingObject Cylinder {
    height 1
    radius 0.5
  }
}